File: simulator.rb
#!/usr/bin/env ruby

require 'singleton'
require_relative 'character'

class Simulator
  include Singleton

  def initialize
    @character_list = []
    @character_list.push(Character.new(20))
    @character_list.push(Character.new)
  end

  def run
    turn = 0
    gameEnd = false
  
    until gameEnd do
      puts "turn #{turn}"
    
      @character_list.each do |character|
        character.update
        character.print_state
        gameEnd = gameEnd || character.dead?
      end

      puts "turn end\n"
    end
  end
end

File: cpp/Simulator.cpp
#include <cstdio>

#include "Simulator.h"


void Simulator::run()
{
    int turn = 0;
    bool gameEnd = false;

    while ( !gameEnd )
    {
        printf("turn : %d\n", turn++);

		for ( auto it = mCharacterVector.begin(); it != mCharacterVector.end(); it++)
		{
            it->update();
            it->printState();

            gameEnd = ( gameEnd || it->isDead());
        }

        printf("turn end\n");

    }
}

void Simulator::init()
{
    mCharacterVector.push_back(Character(20));
    mCharacterVector.push_back(Character());
}


Simulator::Simulator()
{
    init();
}
